openoa/utils/downloader.py

```diff
@@ -0,0 +1,425 @@
+    # save record details to json file
+    outfile = outfile_path+"record_details.json"
+    
+    with open(outfile, "wb") as f:
+        f.write(r.content)
+
+        
+    # download all files
+    files = r_json["files"]
+    for f in files:
+        
+        url_file = f["links"]["self"]
+        
+        file_name = f["key"]
+                
+        outfile = outfile_path+file_name
+        
+        
+        # check if file exists
+        if os.path.exists(outfile):
+            
+            
+            # if it does check the checksum is correct
+            with open(outfile, "rb") as f_check:
+                file_hash = hashlib.md5()
+                while chunk := f_check.read(8192):
+                    file_hash.update(chunk)
+        
+            if f["checksum"][4:]==file_hash.hexdigest():
+                logger.info("File already exists: " + file_name)
+            
+            
+            # download and unzip if the checksum isn't correct
+            else:
+                
+                logger.info("Downloading: " + file_name)
+                logger.info("File size: " + str(round(f["size"]/(1024*1024),2)) + "MB")       
+
+                download_file(url_file,outfile)
+
+                logger.info("Saved to: " + str(outfile) + "\n")
+
+                if outfile.endswith(".zip"):
+                    with ZipFile(outfile) as zipfile:
+                        zipfile.extractall(outfile_path)
+        
+        
+        # download and unzip if the file doesn't exist
+        else:
+            
+            logger.info("\nDownloading: " + file_name)
+            logger.info("File size: " + str(round(f["size"]/(1024*1024),2)) + "MB")       
+
+            download_file(url_file,outfile)
+
+            logger.info("Saved to: " + str(outfile) + "\n")
+
+            if outfile.endswith(".zip"):
+                with ZipFile(outfile) as zipfile:
+                    zipfile.extractall(outfile_path)
+
+
+
+def get_era5(
+        lat,
+        lon,
+        save_pathname,
+        save_filename,
+        ):
+    """
+    Get ERA5 data directly from the CDS service
+    This requires registration on the CDS service
+    See: https://cds.climate.copernicus.eu/api-how-to
+
+    Monthly 10m height data is demonstrated here, as hourly data takes too long to download, 
+    but could be amended and other CDS datasets also used (e.g. CERRA for Europe)
+
+    Args:
+        lat(:obj:`float`): Latitude in WGS 84 spatial reference system (decimal degrees)
+        lon(:obj:`float`): Longitude in WGS 84 spatial reference system (decimal degrees)
+        save_pathname(:obj:`string`): The path where the downloaded reanalysis data will be saved
+        save_filename(:obj:`string`): The file name used to save the downloaded reanalysis data
+
+    Returns:
+        df(:obj:`dataframe`): A dataframe containing time series of the requested reanalysis variables
+        Saved NetCDF annual ERA5 files
+        Saved ERA5 csv file
```

Comment:
These two lines should be added into the docstring description instead of at the end, in addition to where they'll be saved (this will be just be the `save_pathname/save_filename`, but it's generally better to be explicit with documentation).

Reply:
Done for both ERA5 and MERRA2 functions